File: src/main/java/ru/citeck/idea/files/types/citeck/app/EcosApplication.java
package ru.citeck.idea.files.types.citeck.app;

import com.intellij.openapi.diagnostic.Logger;
import ecos.com.fasterxml.jackson210.databind.ObjectMapper;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiFile;
import com.intellij.psi.xml.XmlFile;
import com.intellij.psi.xml.XmlTag;
import com.intellij.psi.xml.XmlTagValue;
import com.intellij.util.io.Compressor;
import lombok.Data;
import org.jetbrains.annotations.Nullable;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.commons.json.YamlUtils;
import ru.citeck.idea.files.types.citeck.EcosArtifact;
import ru.citeck.idea.files.types.citeck.JsonEcosArtifact;
import ru.citeck.idea.files.types.citeck.YamlEcosArtifact;
import ru.citeck.idea.files.types.filters.FileFilter;
import ru.citeck.idea.files.types.filters.FileNameFilter;
import ru.citeck.idea.files.types.filters.FilterAnd;
import ru.citeck.idea.files.types.filters.FilterOr;
import ru.citeck.idea.utils.MavenUtils;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.Optional;
import java.util.zip.Deflater;

public interface EcosApplication extends EcosArtifact {

    Logger log = Logger.getInstance(EcosApplication.class);

    String SOURCE_ID = "eapps/ecosapp";
    String PATH = "/src/main/resources/app/";

    @Override
    default boolean canFetch(PsiFile psiFile) {
        return false;
    }

    @Override
    default byte[] getContent(PsiFile psiFile) {

        try {

            Meta meta = getApplicationMeta(psiFile);
            if (meta == null) {
                throw new RuntimeException("Unable to read application meta");
            }

            String applicationVersion = getApplicationVersion(psiFile);
            if (applicationVersion == null) {
                throw new RuntimeException("Unable to read application version");
            }

            meta.version = applicationVersion;
            Path metaPath = psiFile.getVirtualFile().toNioPath();

            File appDirectory = Optional
                .ofNullable(psiFile.getParent())
                .map(PsiDirectory::getVirtualFile)
                .map(VirtualFile::getPath)
                .map(File::new)
                .orElseThrow(() -> new RuntimeException("Unable to determine application path"));

            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();

            Compressor.Zip zip = new Compressor
                .Zip(byteArrayOutputStream)
                .withLevel(Deflater.BEST_COMPRESSION);

            zip.filter((fileName, path) -> !metaPath.equals(path));
            zip.addDirectory(appDirectory);
            zip.addFile("meta.json", new ObjectMapper().writeValueAsString(meta).getBytes(StandardCharsets.UTF_8));
            zip.close();

            return byteArrayOutputStream.toByteArray();

        } catch (Exception e) {
            throw new RuntimeException(e);
        }

    }

    default @Nullable Meta getApplicationMeta(PsiFile psiFile) {
        try {
            Object data = YamlUtils.read(psiFile.getText());
            if (data == null) {
                return null;
            }
            return Json.getMapper().convertNotNull(data, Meta.class);
        } catch (Exception e) {
            log.error("Application meta reading failed: " + psiFile.getVirtualFile().getPath(), e);
            return null;
        }
    }

    default @Nullable String getApplicationVersion(PsiFile psiFile) {
        return Optional
            .ofNullable(MavenUtils.getPomFile(psiFile))
            .map(XmlFile::getRootTag)
            .map(xmlTag -> xmlTag.findFirstSubTag("version"))
            .map(XmlTag::getValue)
            .map(XmlTagValue::getText)
            .orElse(null);
    }

    @Data
    class Meta {
        private String id;
        private Object name;
        private String version;
    }

    class JSON extends JsonEcosArtifact implements EcosApplication {
        private final FileFilter filter;

        public JSON() {
            super(PATH, SOURCE_ID);
            this.filter = new FilterAnd(super.getFilter(), new FileNameFilter(PATH + "meta.json"));
        }

        @Override
        public FileFilter getFilter() {
            return filter;
        }

        @Override
        public String getMutationAttribute() {
            return "_content";
        }

    }

    class YAML extends YamlEcosArtifact implements EcosApplication {

        private final FileFilter filter;

        public YAML() {
            super(PATH, SOURCE_ID);
            this.filter = new FilterAnd(super.getFilter(), new FilterOr(
                new FileNameFilter(PATH + "meta.yml"),
                new FileNameFilter(PATH + "meta.yaml")
            ));
        }

        @Override
        public FileFilter getFilter() {
            return filter;
        }

        @Override
        public String getMutationAttribute() {
            return "_content";
        }

    }

}

File: src/main/kotlin/ru/citeck/idea/artifacts/type/uncompleted/EcosApplication.kt
package ru.citeck.idea.artifacts.type.uncompleted
/*

import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiFile
import com.intellij.psi.xml.XmlFile
import com.intellij.psi.xml.XmlTag
import com.intellij.psi.xml.XmlTagValue
import com.intellij.util.io.Compressor
import ecos.com.fasterxml.jackson210.databind.ObjectMapper
import lombok.Data
import ru.citeck.ecos.commons.json.Json.mapper
import ru.citeck.ecos.commons.json.YamlUtils.read
import ru.citeck.idea.files.types.citeck.EcosArtifact
import ru.citeck.idea.files.types.citeck.JsonEcosArtifact
import ru.citeck.idea.files.types.citeck.YamlEcosArtifact
import ru.citeck.idea.files.types.filters.FileFilter
import ru.citeck.idea.files.types.filters.FileNameFilter
import ru.citeck.idea.files.types.filters.FilterAnd
import ru.citeck.idea.files.types.filters.FilterOr
import ru.citeck.idea.utils.MavenUtils
import java.io.ByteArrayOutputStream
import java.io.File
import java.nio.charset.StandardCharsets
import java.nio.file.Path
import java.util.*
import java.util.function.Function
import java.util.zip.Deflater

interface EcosApplication : EcosArtifact {
    override fun getContent(psiFile: PsiFile): ByteArray {
        try {
            val meta = getApplicationMeta(psiFile) ?: throw RuntimeException("Unable to read application meta")

            val applicationVersion = getApplicationVersion(psiFile)
                ?: throw RuntimeException("Unable to read application version")

            meta.version = applicationVersion
            val metaPath = psiFile.virtualFile.toNioPath()

            val appDirectory = Optional
                .ofNullable<PsiDirectory>(psiFile.parent)
                .map<VirtualFile> { obj: PsiDirectory -> obj.virtualFile }
                .map<String> { obj: VirtualFile -> obj.path }
                .map<File>(Function<String, File> { pathname: String? -> File(pathname) })
                .orElseThrow<RuntimeException> { RuntimeException("Unable to determine application path") }

            val byteArrayOutputStream = ByteArrayOutputStream()

            val zip = Compressor.Zip(byteArrayOutputStream)
                .withLevel(Deflater.BEST_COMPRESSION)

            zip.filter { fileName: String?, path: Path? -> metaPath != path }
            zip.addDirectory(appDirectory)
            zip.addFile("meta.json", ObjectMapper().writeValueAsString(meta).toByteArray(StandardCharsets.UTF_8))
            zip.close()

            return byteArrayOutputStream.toByteArray()
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    fun getApplicationMeta(psiFile: PsiFile): Meta? {
        try {
            val data = read(psiFile.text) ?: return null
            return mapper.convertNotNull(data, Meta::class.java)
        } catch (e: Exception) {
            log.error("Application meta reading failed: " + psiFile.virtualFile.path, e)
            return null
        }
    }

    fun getApplicationVersion(psiFile: PsiFile?): String? {
        return Optional
            .ofNullable(MavenUtils.getPomFile(psiFile))
            .map { obj: XmlFile -> obj.rootTag }
            .map { xmlTag: XmlTag? -> xmlTag!!.findFirstSubTag("version") }
            .map { obj: XmlTag? -> obj!!.value }
            .map { obj: XmlTagValue -> obj.text }
            .orElse(null)
    }

    @Data
    class Meta {
        private val id: String? = null
        private val name: Any? = null
        internal var version: String? = null
    }

    class JSON : JsonEcosArtifact(PATH, SOURCE_ID), EcosApplication {
        private val filter: FileFilter =
            FilterAnd(super.getFilter(), FileNameFilter(PATH + "meta.json"))

        override fun getFilter(): FileFilter {
            return filter
        }
    }

    class YAML : YamlEcosArtifact(PATH, SOURCE_ID), EcosApplication {
        private val filter: FileFilter = FilterAnd(
            super.getFilter(), FilterOr(
                FileNameFilter(PATH + "meta.yml"),
                FileNameFilter(PATH + "meta.yaml")
            )
        )

        override fun getFilter(): FileFilter {
            return filter
        }
    }

    companion object {
        val log: Logger = Logger.getInstance(
            EcosApplication::class.java
        )

        const val SOURCE_ID: String = "eapps/ecosapp"
        const val PATH: String = "/src/main/resources/app/"
    }
}
*/
